File: src/web/Controllers/CryptosController.cs
using System;
using System.IO;
using System.Threading.Tasks;
using core.fs.Cryptos;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using web.Utils;

namespace web.Controllers
{
    [ApiController]
    [Authorize]
    [Route("api/[controller]")]
    public class CryptosController(Handler service) : ControllerBase
    {
        [HttpGet()]
        public Task<ActionResult> Dashboard() =>
            this.OkOrError(service.Handle(new DashboardQuery(User.Identifier())));

        [HttpGet("{token}")]
        public Task<ActionResult> DetailsAsync([FromRoute] string token) =>
            this.OkOrError(service.Handle(new Details(token)));

        [HttpGet("{token}/ownership")]
        public Task<ActionResult> Ownership([FromRoute]string token) =>
            this.OkOrError(service.Handle(new OwnershipQuery(token, User.Identifier())));

        [HttpDelete("{token}/transactions/{transactionId}")]
        public Task<ActionResult> DeleteTransaction([FromRoute]string token, [FromRoute]Guid transactionId) =>
            this.OkOrError(service.Handle(new DeleteTransaction(token: token, transactionId: transactionId, userId:User.Identifier())));

        [HttpPost("import")]
        public async Task<ActionResult> Import(IFormFile file)
        {
            using var streamReader = new StreamReader(file.OpenReadStream());

            var content = await streamReader.ReadToEndAsync();
            
            var cmd = ImportCryptoCommandFactory.create(file.FileName, content, User.Identifier());

            return await this.OkOrError(service.Handle(cmd));
        }

        [HttpGet("export")]
        public Task<ActionResult> Export() =>
            this.GenerateExport(service.Handle(new Export(User.Identifier())));
    }
}

File: src/web.fs/Controllers/CryptosController.fs
namespace web.Controllers

open System
open System.IO
open System.Threading.Tasks
open Microsoft.AspNetCore.Mvc
open Microsoft.AspNetCore.Authorization
open Microsoft.AspNetCore.Http
open core.fs.Cryptos
open web.Utils

[<ApiController>]
[<Authorize>]
[<Route("api/[controller]")>]
type CryptosController(service: Handler) =
    inherit ControllerBase()

    [<HttpGet>]
    member this.Dashboard() =
        this.OkOrError(service.Handle({DashboardQuery.UserId = this.User.Identifier()}))

    [<HttpGet("{token}")>]
    member this.DetailsAsync([<FromRoute>] token: string) =
        this.OkOrError(service.Handle {Details.Token = token |> core.Cryptos.Token})

    [<HttpGet("{token}/ownership")>]
    member this.Ownership([<FromRoute>] token: string) =
        this.OkOrError(service.Handle {OwnershipQuery.Token = token |> core.Cryptos.Token; UserId = this.User.Identifier()})

    [<HttpDelete("{token}/transactions/{transactionId}")>]
    member this.DeleteTransaction([<FromRoute>] token: string, [<FromRoute>] transactionId: Guid) =
        this.OkOrError(service.Handle {DeleteTransaction.Token = token |> core.Cryptos.Token; TransactionId = transactionId; UserId = this.User.Identifier()})

    [<HttpPost("import")>]
    member this.Import(file: IFormFile) = task {
        use streamReader = new StreamReader(file.OpenReadStream())
        let! content = streamReader.ReadToEndAsync()
        let cmd = ImportCryptoCommandFactory.create file.FileName content (this.User.Identifier())
        return! this.OkOrError(service.Handle cmd)
    }

    [<HttpGet("export")>]
    member this.Export() =
        this.GenerateExport(service.Handle {Export.UserId = this.User.Identifier()})
